lib/streamlit/type_util.py

```diff
@@ -690,6 +776,167 @@ def bytes_to_data_frame(source: bytes) -> DataFrame:
+            # otherwise the values type from infer_dtype would have been empty
+            first_element = next(iter(input_data))
+            if isinstance(first_element, dict):
+                return DataFormat.LIST_OF_RECORDS
+            if isinstance(first_element, (list, tuple, set)):
+                return DataFormat.LIST_OF_ROWS
+    elif isinstance(input_data, dict):
+        if not input_data:
+            return DataFormat.KEY_VALUE_DICT
+        if len(input_data) > 0:
+            first_value = next(iter(input_data.values()))
+            if isinstance(first_value, dict):
+                return DataFormat.COLUMN_INDEX_MAPPING
+            if isinstance(first_value, (list, tuple)):
+                return DataFormat.COLUMN_VALUE_MAPPING
+            if isinstance(first_value, pd.Series):
+                return DataFormat.COLUMN_SERIES_MAPPING
+            # In the future, we could potentially also support the tight & split formats here
+            if is_list_of_scalars(input_data.values()):
+                # Only use the key-value dict format if the values are only scalar values
+                return DataFormat.KEY_VALUE_DICT
+    return DataFormat.UNKNOWN
+
+
+def convert_df_to_data_format(
+    df: pd.DataFrame, data_format: DataFormat
+) -> Union[
+    pd.DataFrame,
+    pd.Index,
+    Styler,
+    pa.Table,
+    np.ndarray[Any, np.dtype[Any]],
+    Tuple[Any],
+    List[Any],
+    Set[Any],
+    Dict[str, Any],
+]:
+    """Convert a dataframe to the specified data format.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        The dataframe to convert.
+
+    data_format : DataFormat
+        The data format to convert to.
+
+    Returns
+    -------
+    pd.DataFrame, pd.Index, Styler, pa.Table, np.ndarray, tuple, list, set, dict
+        The converted dataframe.
+    """
+    if data_format in [
+        DataFormat.EMPTY,
+        DataFormat.PANDAS_DATAFRAME,
+        DataFormat.SNOWPARK_OBJECT,
+        DataFormat.PYSPARK_OBJECT,
+        DataFormat.PANDAS_INDEX,
+        DataFormat.PANDAS_STYLER,
+    ]:
+        return df
+    elif data_format == DataFormat.NUMPY_LIST:
+        # It's a 1-dimensional array, so we only return
+        # the first column as numpy array
+        # Calling to_numpy() on the full DataFrame would result in:
+        # [[1], [2]] instead of [1, 2]
+        return np.array([]) if df.empty else df.iloc[:, 0].to_numpy()
+    elif data_format == DataFormat.NUMPY_MATRIX:
+        return df.to_numpy()
+    elif data_format == DataFormat.PYARROW_TABLE:
+        return pa.Table.from_pandas(df)
+    elif data_format == DataFormat.PANDAS_SERIES:
+        # Select first column in dataframe and create a new series based on the values
+        if len(df.columns) != 1:
+            raise ValueError(
+                f"DataFrame is expected to have a single column but has {len(df.columns)}."
+            )
+        return df[df.columns[0]]
```

Comment:
if len(df.columns) == 0:
  would this break? 
  
I would assume no but im also not a pandas series expert.

Reply:
This would not be possible because of the check above: `if len(df.columns) != 1:` So it can only have one column here.